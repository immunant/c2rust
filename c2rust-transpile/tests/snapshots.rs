use std::env::current_dir;
use std::fs;
use std::path::Path;
use std::process::Command;

use c2rust_transpile::{ReplaceMode, TranspilerConfig};

fn config() -> TranspilerConfig {
    TranspilerConfig {
        dump_untyped_context: false,
        dump_typed_context: false,
        pretty_typed_context: false,
        dump_function_cfgs: false,
        json_function_cfgs: false,
        dump_cfg_liveness: false,
        dump_structures: false,
        verbose: false,
        debug_ast_exporter: false,
        incremental_relooper: true,
        fail_on_multiple: false,
        filter: None,
        debug_relooper_labels: false,
        prefix_function_names: None,
        translate_asm: true,
        use_c_loop_info: true,
        use_c_multiple_info: true,
        simplify_structures: true,
        panic_on_translator_failure: false,
        emit_modules: false,
        fail_on_error: true,
        replace_unsupported_decls: ReplaceMode::Extern,
        translate_valist: true,
        overwrite_existing: true,
        reduce_type_annotations: false,
        reorganize_definitions: false,
        enabled_warnings: Default::default(),
        emit_no_std: false,
        output_dir: None,
        translate_const_macros: false,
        translate_fn_macros: false,
        disable_refactoring: false,
        preserve_unused_functions: false,
        log_level: log::LevelFilter::Warn,
        emit_build_files: false,
        binaries: Vec::new(),
    }
}

fn transpile(platform: Option<&str>, c_path: &Path) {
    let status = Command::new("clang")
        .args(&["-c", "-o", "/dev/null"])
        .arg(c_path)
        .status();
    assert!(status.unwrap().success());

    let (_temp_dir, temp_path) =
        c2rust_transpile::create_temp_compile_commands(&[c_path.to_owned()]);
    c2rust_transpile::transpile(config(), &temp_path, &[]);
    let cwd = current_dir().unwrap();
    let c_path = c_path.strip_prefix(&cwd).unwrap();
    let rs_path = c_path.with_extension("rs");
    let rs = fs::read_to_string(&rs_path).unwrap();
    let debug_expr = format!("cat {}", rs_path.display());

    let name = platform
        .map(|platform| ["transpile", platform].join("-"))
        .unwrap_or("transpile".into());

    insta::assert_snapshot!(name, &rs, &debug_expr);

    let status = Command::new("rustc")
        .args(&["--crate-type", "lib", "--edition", "2021", "-o", "-"])
        .arg(&rs_path)
        .status();
    assert!(status.unwrap().success());
}

#[test]
fn transpile_all() {
    insta::glob!("snapshots/*.c", |x| transpile(None, x));

    // Some things transpile differently on Linux vs. macOS,
    // as they use `unsigned long` and `unsigned long long` differently for builtins.
    // This makes snapshot tests trickier, as the output will be OS-dependent.
    // We handle this by adding OS name to the snapshot result filename.
    #[allow(unused)]
    let platform = "unknown";

    #[cfg(target_os = "linux")]
    let platform = "linux";
    #[cfg(target_os = "macos")]
    let platform = "macos";

    insta::with_settings!({snapshot_suffix => platform}, {
        insta::glob!("snapshots/platform-specific/*.c", |x| transpile(Some(platform), x));
    });
}
