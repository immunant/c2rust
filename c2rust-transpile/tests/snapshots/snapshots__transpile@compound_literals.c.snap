---
source: c2rust-transpile/tests/snapshots.rs
expression: cat tests/snapshots/compound_literals.rs
input_file: c2rust-transpile/tests/snapshots/compound_literals.c
---
#![allow(
    dead_code,
    non_camel_case_types,
    non_snake_case,
    non_upper_case_globals,
    unused_assignments,
    unused_mut
)]
#![feature(raw_ref_op)]
#[no_mangle]
pub static mut static_single_int: ::core::ffi::c_int = 42;
#[no_mangle]
pub static mut static_single_int_ptr: *mut ::core::ffi::c_int =
    &42 as *const ::core::ffi::c_int as *mut ::core::ffi::c_int;
#[no_mangle]
pub static mut static_int_ptr_to_array: *mut ::core::ffi::c_int = [42, 9001].as_ptr() as *mut _;
#[no_mangle]
pub static mut static_int_array_ptr: *mut [::core::ffi::c_int; 2] =
    &[42, 9001] as *const [::core::ffi::c_int; 2] as *mut [::core::ffi::c_int; 2];
pub const SINGLE_INT: ::core::ffi::c_int = 42 as ::core::ffi::c_int;
pub const INT_ARRAY: [::core::ffi::c_int; 2] =
    [42 as ::core::ffi::c_int, 9001 as ::core::ffi::c_int];
pub const CHAR_ARRAY: [::core::ffi::c_char; 6] =
    unsafe { ::core::mem::transmute::<[u8; 6], [::core::ffi::c_char; 6]>(*b"hello\0") };
#[no_mangle]
pub unsafe extern "C" fn local_compound_literals() {
    let mut single_int: ::core::ffi::c_int = 42 as ::core::ffi::c_int;
    let mut fresh0: ::core::ffi::c_int = 42 as ::core::ffi::c_int;
    let mut single_int_ptr: *mut ::core::ffi::c_int = &raw mut fresh0;
    let mut fresh1: [::core::ffi::c_int; 2] =
        [42 as ::core::ffi::c_int, 9001 as ::core::ffi::c_int];
    let mut int_ptr_to_array: *mut ::core::ffi::c_int = &raw mut fresh1 as *mut ::core::ffi::c_int;
    let mut fresh2: [::core::ffi::c_char; 6] =
        ::core::mem::transmute::<[u8; 6], [::core::ffi::c_char; 6]>(*b"hello\0");
    let mut char_ptr_to_array: *mut ::core::ffi::c_char =
        &raw mut fresh2 as *mut ::core::ffi::c_char;
    let mut fresh3: [::core::ffi::c_int; 2] =
        [42 as ::core::ffi::c_int, 9001 as ::core::ffi::c_int];
    let mut int_array_ptr: *mut [::core::ffi::c_int; 2] = &raw mut fresh3;
    let mut fresh4: [::core::ffi::c_char; 6] =
        ::core::mem::transmute::<[u8; 6], [::core::ffi::c_char; 6]>(*b"hello\0");
    let mut char_array_ptr: *mut [::core::ffi::c_char; 6] = &raw mut fresh4;
    let mut macro_single_int: ::core::ffi::c_int = SINGLE_INT;
    let mut macro_single_int_ptr: *mut ::core::ffi::c_int =
        &mut SINGLE_INT as *mut ::core::ffi::c_int;
    let mut macro_int_ptr_to_array: *mut ::core::ffi::c_int = INT_ARRAY.as_mut_ptr();
    let mut macro_char_ptr_to_array: *mut ::core::ffi::c_char = CHAR_ARRAY.as_mut_ptr();
    let mut macro_int_array_ptr: *mut [::core::ffi::c_int; 2] =
        &mut INT_ARRAY as *mut [::core::ffi::c_int; 2];
    let mut macro_char_array_ptr: *mut [::core::ffi::c_char; 6] =
        &mut CHAR_ARRAY as *mut [::core::ffi::c_char; 6];
}
